File: js/Cjk_test.js
goog.require('goog.testing.asserts');
goog.require('goog.testing.jsunit');
goog.require('twim.CjkDecoder');
goog.require('twim.CjkEncoder');

/**
 * @param{!Uint8Array} bytes
 */
function checkRoundtrip(bytes) {
  let codePoints = twim.CjkEncoder.encode(bytes);
  let decoded = twim.CjkDecoder.decode(codePoints);
  assertSameElements(bytes, decoded);
}

function testEmpty() {
  checkRoundtrip(new Uint8Array([]));
}

function testOneByte() {
  checkRoundtrip(new Uint8Array([42]));
}

function testZeroTruncation() {
  let codePoints = twim.CjkEncoder.encode(new Uint8Array([42, 0, 0, 0, 0, 0, 43, 0, 0, 0]));
  let decoded = twim.CjkDecoder.decode(codePoints);
  assertSameElements(new Uint8Array([42, 0, 0, 0, 0, 0, 43]), decoded);
}

function testRandom() {
  checkRoundtrip(new Uint8Array([
    0x20, 0x4c, 0x75, 0x2f, 0xa1, 0xd6, 0xe3, 0xb4, 0x21, 0x1f, 0x43,
    0x77, 0x1d, 0xc8, 0x4b, 0x7c, 0xfa, 0x6d, 0xf6, 0x6c, 0x8a, 0x78,
    0x9c, 0x47, 0x60, 0xdd, 0xae, 0x90, 0xb8, 0xee, 0x38, 0xb5, 0x07,
    0x9e, 0x2a, 0x3d, 0xb0, 0xc0, 0x4e, 0x69, 0xcf, 0x32
  ]));
}

File: java/ru/eustas/twim/CjkTest.java
package ru.eustas.twim;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import static org.junit.Assert.assertArrayEquals;

@RunWith(JUnit4.class)
public class CjkTest {
  private void checkRoundtrip(byte[] bytes) {
    int[] codePoints = CjkEncoder.encode(bytes);
    byte[] decoded = CjkDecoder.decode(codePoints);
    assertArrayEquals(bytes, decoded);
  }

  @Test
  public void testCoverage() {
    new CjkDecoder();
    new CjkEncoder();
    new CjkTransform();
  }

  @Test
  public void testEmpty() {
    checkRoundtrip(new byte[0]);
  }

  @Test
  public void testOneByte() {
    checkRoundtrip(new byte[] {42});
  }

  @Test
  public void testZeroTruncation() {
    int[] codePoints = CjkEncoder.encode(new byte[] {42, 0, 0, 0, 0, 0, 43, 0, 0, 0});
    byte[] decoded = CjkDecoder.decode(codePoints);
    assertArrayEquals(new byte[] {42, 0, 0, 0, 0, 0, 43}, decoded);
  }

  @Test
  public void testRandom() {
    checkRoundtrip(new byte[] {(byte) 0x20, (byte) 0x4c, (byte) 0x75, (byte) 0x2f, (byte) 0xa1,
        (byte) 0xd6, (byte) 0xe3, (byte) 0xb4, (byte) 0x21, (byte) 0x1f, (byte) 0x43, (byte) 0x77,
        (byte) 0x1d, (byte) 0xc8, (byte) 0x4b, (byte) 0x7c, (byte) 0xfa, (byte) 0x6d, (byte) 0xf6,
        (byte) 0x6c, (byte) 0x8a, (byte) 0x78, (byte) 0x9c, (byte) 0x47, (byte) 0x60, (byte) 0xdd,
        (byte) 0xae, (byte) 0x90, (byte) 0xb8, (byte) 0xee, (byte) 0x38, (byte) 0xb5, (byte) 0x07,
        (byte) 0x9e, (byte) 0x2a, (byte) 0x3d, (byte) 0xb0, (byte) 0xc0, (byte) 0x4e, (byte) 0x69,
        (byte) 0xcf, (byte) 0x32});
  }
}
